auth/src/main/java/org/aerogear/auth/impl/OIDCAuthCodeImpl.java

```diff
@@ -20,6 +43,112 @@ public OIDCAuthCodeImpl(final AuthServiceConfig config) {
      */
     @Override
     public Principal authenticate(final ICredential credential) {
-        throw new IllegalStateException("Not implemented");
+        UserPrincipalImpl user = null;
+        try {
+            userIdentity = getIdentityInformation(credential);
+            user = UserPrincipalImpl
+                    .newUser()
+                    .withAuthenticator(this)
+                    .withUsername(parseUsername())
+                    .withCredentials(credential)
+                    .withEmail(parseEmail())
+                    .withRoles(parseRoles())
+                    .build();
+        } catch (JSONException e) {
+            e.printStackTrace();
+        }
+    return user;
+    }
+
+    private String parseUsername() throws JSONException {
+        String username = "Unknown Username";
+        if (userIdentity != null) {
+            // get the users username
+            if (userIdentity.has(USERNAME) && userIdentity.getString(USERNAME).length() > 0) {
+                username = userIdentity.getString(USERNAME);
+            }
+        }
+        return username;
+    }
+
+    private String parseEmail() throws JSONException {
+        String emailAddress = "Unknown Email";
+        if (userIdentity != null) {
+            // get the users email
+            if (userIdentity.has(EMAIL) && userIdentity.getString(EMAIL).length() > 0) {
+                emailAddress = userIdentity.getString(EMAIL);
+            }
+        }
+        return emailAddress;
+    }
+
+    private List<IRole> parseRoles() throws JSONException {
+        List<IRole> roles = new ArrayList();
+        if (userIdentity != null) {
+            List<RealmRole> realmRoles = parseRealmRoles();
+            if (realmRoles != null) {
+                roles.addAll(realmRoles);
+            }
+            List<ClientRole> clientRoles = parseClientRoles();
+            if (clientRoles != null) {
+                roles.addAll(clientRoles);
+            }
+        }
+        return roles;
+    }
+
+    private List<RealmRole> parseRealmRoles() throws JSONException {
+        List realmRoles = new ArrayList<RealmRole>();
+        if (userIdentity.has(REALM) && userIdentity.getJSONObject(REALM).has(ROLES)) {
+            String tokenRealmRolesJSON = userIdentity.getJSONObject(REALM).getString(ROLES);
+
+            String realmRolesString = tokenRealmRolesJSON.substring(1, tokenRealmRolesJSON.length() - 1); //remove the array brackets
+            String roles[] = realmRolesString.split(COMMA);
+            realmRoles = Arrays.asList(roles);
+        }
+        return realmRoles;
+    }
+
+    private List<ClientRole> parseClientRoles() throws JSONException {
```

Comment:
I would suggest returning a collection instead of a List, so that we can freely change the implementation without affecting the caller